### PR TITLE
Uplift `Relate`/`TypeRelation` into `rustc_next_trait_solver`

### PR DESCRIPTION
**Pull request by compiler-errors**
For use in the new solver. This doesn't yet uplift `ObligationEmittingRelation`.

r? lcnr

### CODE REVIEW

**Comment by compiler-errors** [comment]
One structure question: Currently this uplifts `TypeRelation` and `Relate` into `rustc_next_trait_solver`. This is actually somewhat problematic, since I would eventually like to make `InferCtxtLike` have a `eq<T: Relate<Self>>()` method, but `InferCtxtLike` is defined in `rustc_type_ir`.
    * We may need to move `InferCtxtLike` down `rustc_next_trait_solver` (which affects the ability to use `DebugWithInfcx`) or move relations up into `rustc_type_ir`.

File: compiler/rustc_type_ir/src/const_kind.rs [hunk 1]
```diff
@@ -204,3 +204,36 @@ impl<CTX> HashStable<CTX> for InferConst {
         }
     }
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "nightly", derive(HashStable_NoContext, TyEncodable, TyDecodable))]
+pub enum BoundConstness {
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/const_kind.rs:210]: idk how I feel about `BoundConstness` being part of `const_kind`, maybe move it to `const_trait` module or sth?

File: compiler/rustc_type_ir/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    FloatMismatch(ExpectedFound<ty::FloatTy>),
+    Traits(ExpectedFound<I::DefId>),
+    VariadicMismatch(ExpectedFound<bool>),
+
+    /// Instantiating a type variable with the given type would have
+    /// created a cycle (because it appears somewhere within that
+    /// type).
+    CyclicTy(I::Ty),
+    CyclicConst(I::Const),
+    ProjectionMismatched(ExpectedFound<I::DefId>),
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/error.rs:62]:
```suggestion
    ProjectionMismatch(ExpectedFound<I::DefId>),
```
idk if you want to fix it in this PR

File: compiler/rustc_type_ir/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    Clone(bound = ""),
+    Copy(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = ""),
+    Debug(bound = "")
+)]
+#[derive(TypeVisitable_Generic)]
+#[rustc_pass_by_value]
+pub enum TypeError<I: Interner> {
+    Mismatch,
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/error.rs:35]: not for this PR

```suggestion
    Misc,
```

File: compiler/rustc_type_ir/src/relate.rs [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+
+pub trait TypeRelation<Infcx, I = <Infcx as InferCtxtLike>::Interner>: Sized
+where
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
+{
+    fn tcx(&self) -> I;
+
+    /// Returns a static string we can use for printouts.
+    fn tag(&self) -> &'static str;
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/relate.rs:20]: not necessarily for this PR. Remove this and just use `std::any::type_name` for debug printing instead

File: compiler/rustc_type_ir/src/relate.rs [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
+    R: TypeRelation<Infcx>,
+>(
+    relation: &mut R,
+    a_arg: I::GenericArgs,
+    b_arg: I::GenericArgs,
+) -> RelateResult<I, I::GenericArgs>
+where
+    I::GenericArg: Relate<Infcx>,
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_type_ir/src/relate.rs:101]:
```rust
trait Sup {
    type Assoc;
}
trait Trait: Sup<Assoc: Copy> {}

fn foo<T: Trait>(x: T::Assoc) -> (T::Assoc, T::Assoc) {
    (x, x)
}
```
tbh I am actually surpised that that works, don't fully understand why it does

Can add
```rust
trait Interner {
    type GenericArg;
}
// To allow an impl in `rustc_infer` while still having super traits
trait InternerForInfcx<Infcx: ?Sized>: Interner<GenericArg: Relate<Infcx>> {}

trait InferCtxtLike {
    type Interner: InternerForInfcx<Self>;
}


fn foo<Infcx: InferCtxtLike<Interner = I>, I: InternerForInfcx<Infcx>>() {}
```
:3

**Review by lcnr — COMMENTED** [review]
stopped reviewing after getting to the comment in `relate.rs`

as a general note: would you mind moving some of the param bounds into where-bounds, I personally find multi-line generic param listings hard to read

**Comment by compiler-errors** [comment]
I think this is basically ready for another round of review

@rustbot ready

File: compiler/rustc_middle/src/ty/error.rs [hunk 6]
```diff
@@ -1,91 +1,26 @@
...
-/// Explains the source of a type err in a short, human readable way. This is meant to be placed
-/// in parentheses after some larger message. You should also invoke `note_and_explain_type_err()`
-/// afterwards to present additional details, particularly when it comes to lifetime-related
-/// errors.
+#[extension(pub trait TypeErrorToStringExt<'tcx>)]
+/// Explains the source of a type err in a short, human readable way.
+/// This is meant to be placed in parentheses after some larger message.
+/// You should also invoke `note_and_explain_type_err()` afterwards
+/// to present additional details, particularly when it comes to lifetime-
+/// related errors.
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/error.rs:21]:
```suggestion
/// Explains the source of a type err in a short, human readable way.
/// This is meant to be placed in parentheses after some larger message.
/// You should also invoke `note_and_explain_type_err()` afterwards
/// to present additional details, particularly when it comes to lifetime-
/// related errors.
#[extension(pub trait TypeErrorToStringExt<'tcx>)]
```

File: compiler/rustc_type_ir/src/ty_kind.rs [hunk 7]
```diff
@@ -1070,3 +1070,48 @@ impl<I: Interner> DebugWithInfcx<I> for FnSig<I> {
...
+#[derive(derivative::Derivative)]
+#[derivative(
+    Copy(bound = ""),
+    Clone(bound = ""),
+    Debug(bound = ""),
+    Default(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = "")
+)]
+pub enum VarianceDiagInfo<I: Interner> {
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/ty_kind.rs:1090]: feel like this belongs into `relate.rs` instead? :thinking:

**Review by lcnr — COMMENTED** [review]
r=me after nits

File: compiler/rustc_type_ir/src/ty_kind.rs (hunk 7 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/ty_kind.rs:1090]: why the heck did I put it there lol

**Comment by lcnr** [comment]
going to wait on #125958 with this (if we're able to merge that today)

**Comment by compiler-errors** [comment]
Yeah I already locally rebased it onto that PR expecting it would land first, will push the rebased one when it's done.

**Pull request merged**